bigquery/google/cloud/bigquery/retry.py

```diff
@@ -27,6 +31,50 @@
 )
 
 
+class _TimeoutableHTTPAdapter(requests.adapters.HTTPAdapter):
```

Comment:
This feels like something we could put in `google-cloud-core`. I imagine it'd be relevant for our other REST-based clients.

CC @crwilcox @busunkim96

Reply:
Also, could you explain why this is necessary? It appears as though requests has a `timeout` parameter already. https://2.python-requests.org//en/latest/api/#requests.request

Reply:
TL; DR - it allows the timeout customization to be made from BigQuery itself, i.e. without further modifying the core, which initially seemed a good idea.

---

Deep down inside the API Core's [_http. JSONConnection._do_request()](https://github.com/googleapis/google-cloud-python/blob/b7ba9183a5384839e05afc9d26ddc4e5619ab3bf/core/google/cloud/_http.py#L268-L295), `request()` is invoked without a timeout. This `Session.request()` method then invokes [Session.send()](https://github.com/psf/requests/blob/d2590ee46c0641958b6d4792a206bd5171cb247d/requests/sessions.py#L530), which in turn obtains the HTTP adapter and [calls its send()](https://github.com/psf/requests/blob/d2590ee46c0641958b6d4792a206bd5171cb247d/requests/sessions.py#L636-L643) method with all kwargs (by default, the `timeout` is `None` in the default HTTPAdapter's [send() method](https://github.com/psf/requests/blob/d2590ee46c0641958b6d4792a206bd5171cb247d/requests/adapters.py#L394)).

Injecting a custom `timeout` seemed more easy by directly customizing the adapter from the BigQuery client library itself, rather than adding a new `timeout` parameter to several core methods and dragging it through the call stack:

**(BigQuery)**
- `Client._get_query_results()` (already accepts timeout)
- `Client._call_api()`

(retry layers omitted)

**(API core)**
- `_http.JSONConnection.api_request()`
- `_http.JSONConnection._make_request()`
- `_http.JSONConnection._do_request()`

**(requests)**
- `google.auth.transport.requests.AuthorizedSession.request()` (already accepts `**kwargs`)
- `sessions.Session.request()` (already accepts timeout)

Additionally, the same change would have to be made to the [credentials.before_request()](https://github.com/googleapis/google-auth-library-python/blob/0c33e9c0fe4f87fa46c8f1a5afe725a467ac5fcc/google/auth/transport/requests.py#L211-L213) call that eventually [requests a token](https://github.com/googleapis/google-auth-library-python/blob/08272d89667a901d0dff6c4ba53d5b30fcc29e81/google/oauth2/_client.py#L105-L106) through the `google.oauh2` client (again without an explicit timeout currently).

---


Now, if this `timeout` is something we want to add to API Core and provide to other libraries as well, then updating a few API Core `_http.JSONConnection` methods (and possibly the `oauth2` client) is probably better long-term - after all, passing an optional `timeout` parameter is cleaner than having to wrap `self._call_api()` calls into a context manager.

Reply:
> Additionally, the same change would have to be made to the credentials.before_request() call that eventually requests a token through the google.oauh2 client (again without an explicit timeout currently).

Given that we've seen annoying "hang forever" behavior with network hiccups, this is worth doing across all our libraries, including the auth layer.

Reply:
@tswast Can the P1 issue with `QueryJob.done()` method wait for a bit more in favor of a proper fix to API core?

(the quick fix proposed here mitigates the symptom without changes to core, provided that we are fine with temporarily not supporting the "strict" Retry mode)

Reply:
To summarize, these would be the changes:
- Change the behavior of core.Retry to raise a timeout error at (approximately) the retry deadline (#9873)
- Add `timeout` parameter to `_http.JSONConnection.api_request()`
- Add `timeout` parameter to `google.auth.crendetials.Credentials.before_request()` and  `google.auth.crendetials.Credentials.refresh()`. The same for Credentials' subclasses, e.g. `google.oauth2.service_account.Credentials`
- Provide the new `timeout` parameter to all `bigquery.Client._call_api()` calls in various public methods.